### PR TITLE
Fix/audit

### LINKED ISSUE #156: Boolean conditions in require statements can be simplified

### LINKED ISSUE #157: Naming conflict with transferMatchingFunds in FundsManager and GrantFactory
rename to `_transferMatchingFunds` (as it's internal) and change in `QFI`

### LINKED ISSUE #158: Anyone can pass mismatched length encryption keys and message batches to PublishMessageBatch
Add require statement to check that the two arrays' length match.

Added check and test passing mismatched arrays.

### LINKED ISSUE #159: Incorrect error messages
Changed messages and added tests

fix/incorrect-error-messages
commit: da58fb28ab9a872e73cd325d0524bf232f06ea2c

### LINKED ISSUE #160: Miscellaneous mistakes

### LINKED ISSUE #162: Potential for Re-entrancy in QFI

### LINKED ISSUE #163: Transactions do not revert when there aren’t enough funds
Added require statement to check for enough funds.

### LINKED ISSUE #164: Doesn’t take into account ERC20s that are deflationary or charge a fee upon transfer

### LINKED ISSUE #165: Missing re-initialization
Implemented changes to reflect contributions per round.

### LINKED ISSUE #166: Doesn’t take into account ERC20s with blacklisting abilities
Updated docs and added note to exercise caution when choosing a token.

### PR DESCRIPTION
**Pull request by daodesigner**

### CODE REVIEW

File: .husky/pre-commit [hunk 1]
```diff
@@ -1,37 +1,6 @@
 #!/bin/sh
 
+BASE_DIR="$( cd "$( dirname "$0" )" && pwd )"
 
-# Return `true` if files under `packages/contracts/contracts/` directory has been changed or added. Otherwise, return `false`
```
Comment (daodesigner) [inline .husky/pre-commit:4]: keep this section

**Review by daodesigner — COMMENTED** [review]
Changes to contracts are neat and and appreciate you taking care of nits like _ for internal functions, commenting the code, and updating the natspec

Missing integration tests (please do them on a separate file, ie. tests/QF/09-audit)

File: .husky/pre-commit (hunk 1 above)
Comment (ctrlc03) [reply .husky/pre-commit:4]: This is taken from here: https://github.com/quadratic-funding/qfi/commit/af01221850a843e33f3c7f171c0ae637b58d8a15 - what should I change? @daodesigner

**Comment by baumstern** [comment]
🚀

File: .husky/pre-commit (hunk 1 above)
Comment (daodesigner) [reply .husky/pre-commit:4]: oo nvm then

**Pull request merged**